drivers/spi/spi_sam0.c

```diff
@@ -0,0 +1,432 @@
+		ctrla.bit.DIPO = 2;
+	} else {
+		ctrla.bit.DIPO = 0;
+	}
+
+	ctrla.bit.ENABLE = 1;
+	ctrlb.bit.RXEN = 1;
+
+	if (SPI_WORD_SIZE_GET(config->operation) != 8) {
+		return -ENOTSUP;
+	}
+
+	/* 8 bits per transfer */
+	ctrlb.bit.CHSIZE = 0;
+
+	/* Use the requested or next higest possible frequency */
+	div = (SOC_ATMEL_SAM0_GCLK0_FREQ_HZ / config->frequency) / 2 - 1;
+	div = max(0, min(UINT8_MAX, div));
+
+	/* Update the configuration only if it has changed */
+	if (regs->CTRLA.reg != ctrla.reg || regs->CTRLB.reg != ctrlb.reg ||
+	    regs->BAUD.reg != div) {
+		regs->CTRLA.bit.ENABLE = 0;
+		wait_synchronization(regs);
+
+		regs->CTRLB = ctrlb;
+		wait_synchronization(regs);
+		regs->BAUD.reg = div;
+		wait_synchronization(regs);
+		regs->CTRLA = ctrla;
+		wait_synchronization(regs);
+	}
+
+	return 0;
+}
+
+static bool spi_sam0_transfer_ongoing(struct spi_sam0_data *data)
+{
+	return spi_context_tx_on(&data->ctx) || spi_context_rx_on(&data->ctx);
+}
+
+static void spi_sam0_shift_master(SercomSpi *regs, struct spi_sam0_data *data)
+{
+	u8_t tx;
+	u8_t rx;
+
+	if (spi_context_tx_on(&data->ctx)) {
+		tx = *(u8_t *)(data->ctx.tx_buf);
+	} else {
+		tx = 0;
+	}
+
+	while (!regs->INTFLAG.bit.DRE) {
+	}
+
+	regs->DATA.reg = tx;
+	spi_context_update_tx(&data->ctx, 1, 1);
+
+	while (!regs->INTFLAG.bit.RXC) {
+	}
+
+	rx = regs->DATA.reg;
+
+	if (spi_context_rx_on(&data->ctx)) {
+		*data->ctx.rx_buf = rx;
+		spi_context_update_rx(&data->ctx, 1, 1);
+	}
+}
+
+/* Fast path that transmits a buf */
+static void spi_sam0_fast_tx(SercomSpi *regs, const struct spi_buf *tx_buf)
+{
+	const u8_t *p = tx_buf->buf;
+	const u8_t *pend = tx_buf->buf + tx_buf->len;
+	u8_t ch;
+
+	while (p != pend) {
+		ch = *p++;
+		while (!regs->INTFLAG.bit.DRE) {
+		}
+		regs->DATA.reg = ch;
+	}
+
+	/* Note that the RX buf is full and the transmit may be ongoing */
+}
+
+/* Fast path that reads into a buf */
+static void spi_sam0_fast_rx(SercomSpi *regs, struct spi_buf *rx_buf)
+{
+	u8_t *p = rx_buf->buf;
+	size_t len = rx_buf->len;
+
+	while (regs->INTFLAG.bit.RXC) {
+		(void)regs->DATA.reg;
+	}
+
+	if (len <= 0) {
+		return;
+	}
+
+	/* Load the first outgoing byte */
+	while (!regs->INTFLAG.bit.DRE) {
+	}
+	regs->DATA.reg = 0;
+
+	while (len--) {
+		if (len != 0) {
+			while (!regs->INTFLAG.bit.DRE) {
+			}
+			regs->DATA.reg = 0;
+		}
+
+		while (!regs->INTFLAG.bit.RXC) {
+		}
+
+		*p++ = regs->DATA.reg;
+	}
+
+	/* Note that all transmits are complete and the RX buf is empty */
+}
+
+/* Fast path that writes and reads bufs of the same length */
+static void spi_sam0_fast_txrx(SercomSpi *regs, const struct spi_buf *tx_buf,
+			       struct spi_buf *rx_buf)
+{
+	const u8_t *psrc = tx_buf->buf;
+	u8_t *p = rx_buf->buf;
+	size_t len = rx_buf->len;
+
+	while (regs->INTFLAG.bit.RXC) {
+		(void)regs->DATA.reg;
+	}
+
+	if (len <= 0) {
+		return;
+	}
+
+	/* Load the first outgoing byte */
+	while (!regs->INTFLAG.bit.DRE) {
+	}
+	regs->DATA.reg = *psrc++;
+
+	while (len--) {
```

Comment:
do the same as previous function.

Reply:
Ack.  Depends on review of the previous function.